### ISSUE TITLE
Snippet Duplication (3 hrs)

### ISSUE THREAD

**Issue by cmiya** [cwrc/HuViz#9]
Problem:
- Duplicate snippets. We can see the text snippets associated with each edge, but they duplicate or triplicate.

Solution:
- Eliminate all snippets that have in their path DIV 1 DIV 2 and DIV 0. There will be edges that have zero snippets connected to them.
- Need to get rid of duplication from our end: John will fix triples. (Ensure Div0 is not there. If Div2 then no Div 1, if no DIV2 then DIV2).

**Comment by ilovan** [comment]
My guess is that we will have a better idea of how to proceed with this once #41 is done, right @smurp?
I am asking in terms of scoping this work piece in the new contract.

**Comment by smurp** [comment]
Yes, the collective suspicion was that not including DIV0 or DIV1 snippets
would solve much of this problem

On Tue, Oct 13, 2015 at 2:14 PM, Mihaela Ilovan <notifications@github.com>
wrote:

> My guess is that we will have a better idea of how to proceed with this
> once #41 <https://github.com/cwrc/HuViz/issues/41> is done, right @smurp
> <https://github.com/smurp>?
> I am asking in terms of scoping this work piece in the new contract.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cwrc/HuViz/issues/9#issuecomment-147839271>.
>



-- 
Shawn Murphy
smurp@smurp.com
(780) 903-2428
http://www.smurp.com

**Comment by smurp** [comment]
I've suppressed snippets ending with DIV0--1.  I think that does the trick.

**Comment by smurp** [comment]
See what it looks like without the DIV0--1 snippets over at: http://beta.huviz.dev.nooron.com/

**Comment by susanirenebrown** [comment]
Definitely an improvement but we are still getting duplicates.  

For instance if there is a div 1 or 2 with a title and then a paragraph, we are getting duplicates of the snipped with and without title. We really need to try to get down to the smallest surrounding unit, which will usually be P, CHRONSTRUCTUCT, DATASTRUCT, etc. and ignore the larger tags. That will also help restrict the length of snippets which are currently way too long to allow the reader to find the relevant text easily.

**Comment by cmiya** [comment]
Yes, I'm also noticing the doubling of paragraphs with/without title. Other instances of doubling don't seem to have a connection to the title. For example, in the Ballantyne dataset there are three completely identical snippets with the labels: b--ballrm--0--DIV1--4, b--ballrm--0--P--3, b--ballrm--0--DIV2--3.

#### Other observations about the snippets:
- We need a way to distinguish the citations from the regular text.
- It would help for readability to eliminate the spaces before periods and commas.
- As mentioned in #50, I think we need subheadings. When multiple windows open up, the headers are not much help if they all have the same title (ex: seven windows pop up, all titled Margaret Atwood). Having subheadings would help to sort through them AND to contextualize the body text, which sometimes doesn't make sense on its own.

**Comment by cmiya** [comment]
@smurp Any progress on the lingering duplication issues?

**Comment by smurp** [comment]
## Causes

**Snippet duplication** is a misnomer with a better term for what is happening here being **snippet equivalence** because there is not an error occurring, what is happening is that there are invisible XML differences in between the different nested snippets.

To see the distinction, let's look at what is happening in these two cases (you have to use Chrome and you have to paste these strings including the `view-source:` part into the location bar of a new tab to see the source XML)

* `view-source:http://alpha.huviz.dev.nooron.com/snippet/orlando/b--ballrm--0--DIV1--4/`

In this case 
* `b--ballrm--0--DIV1--4`
* `b--ballrm--0--P--3`
* `b--ballrm--0--DIV2--3`

all contain the same text and hence can be accused of *snippet duplication* and this would argue in favour of us suppressing all `DIV1*` snippets, but...

* `view-source:http://alpha.huviz.dev.nooron.com/snippet/orlando/b--ballrm--0--DIV1--3/`

In this case the `b--ballrm--0--DIV1--3` part contains text that is not visible at deeper levels ie `b--ballrm--0--DIV2--2` and hence we would not experience **snippet equivalence**.

## Solutions

1. Suppress any `DIV1` edges as we are doing with `DIV0` edges

  * **Financial Cost** 1 hr of coding or less
  * **Failure** will lose some data as in the example above, see [/b--ballrm--0--DIV1--3](http://alpha.huviz.dev.nooron.com/snippet/orlando/b--ballrm--0--DIV1--3/) ie we will erroneously suppress some edges which are not really textually equivalent to any other, `b--ballrm--0--DIV2--2` in this case is similar but distinct

1. Really, **very** complicated work inside orlandoScrape to figure out (whenever a newly seen snippet_id comes along) whether it would display **snippet equivalence** with any others which have already been encountered in the current extraction.  Unclear in this algorithm is what to do with situations where textually equivalent snippets are being encountered in contextually different situations, ie predicate A points at one snippet and predicate B points at a textually equivalent snippet -- what should we do?

    * **Financial Cost**: unknown, at least 5 hrs maybe 10); 
    * **Computational Cost**: it would probably slow orlandoScrape down by a factor of 3 and it is already painful
    * **Feature**: the _**snippet counts on edge labels**_ will still be accurate

1. Check for **snippet equivalence** on the client side instead and suppress the displaying of snippet windows which are textually equivalent to others which have already been displayed.  Unclear in this approach is how to deal with situations where different predicates point to the different snippets which happen to be textually equivalent.

  * **Financial Cost**: unknown, at least 5 hours, see **Failure** below
  * **Computational Cost**: minimal
  * **Failure**: the _**snippet counts on edge labels**_ will be wrong and not repairable until snippets get displayed, whereupon those counts could be rendered accurate at the cost of several hours more programming work

1. Live with it 

  * **Feature** it's free! and an unpredictable amount of work is avoided for all of us
  * **Failure** no change

## Further Notes

1. These phenomena are a consequence of our current XML-based and `orlandoScrape.py`-based ways of extracting information from the Orlando corpus and have no impact on any other use of HuViz, so are strictly a HuViz-as-applied-to-Orlando-at-the-moment issue. Not to downplay the importance of that, but just to contextualize it.
1. This comment took 2+ hrs to prepare

**Comment by cmiya** [comment]
I see what you mean. I think at this point, our labour could be better used elsewhere. We should postpone making a decision on this until after a meeting with @SusanBrown and the rest of the HuViz team. For now, I say we remove this from the testing milestone and add a "question" and "improvement" tag.

**Comment by cmiya** [comment]
@SusanBrown Any thoughts on postponing this issue (and #10) until after testing?

**Comment by smurp** [comment]
As far as I understand we are acknowledging that the next steps along this path remain to be decided.

**Comment by cmiya** [comment]
@smurp Does this issue no longer apply? Does it need to be closed... or?

**Comment by susanirenebrown** [comment]
I think this got largely addressed already but it will be moot when we do the new extraction of data so suggest that we can close this one.
